### PR TITLE
feature: x-omitempty extension for general types

### LINKED ISSUE #15: validation: unique scopes in security definitions
each security scope in a security definition should be unique

At the top level of the swagger object there is a property SecurityDefinitions: https://github.com/go-swagger/go-swagger/blob/master/spec/swagger.go#L48  
This object is a map https://github.com/go-swagger/go-swagger/blob/master/spec/swagger.go#L169-L170

The value of that property is a map, to resolve this issue you can't work with the actual swagger object.
You have to work with the map that's obtained by parsing the raw byte value of the spec to a map[string]interface{}

The way the object model is set up is that it uses a map, by definition keys are unique in a map, so once parsed into the swagger object model this error would be lost.

you have to look into securityDefintions[key].scopes for unique values. The values only have to be unique within a particular key.

### PR DESCRIPTION
**Pull request by zhuangqh**
### Ⅰ. Describe what this PR did

support `x-omityempty` extension for general type. 
Now we can use `x-omityempty` to control presence of omitempty tag to be used by JSON Marshaler.

By default, `x-omitempty` is false for array type, true for other.

### Ⅱ. Does this pull request fix one issue?

#1623 

### III. Describe how to verify it

I created a test for this feature.

```yaml
---
swagger: '2.0'
info:
  version: 0.0.0
  title: Breaking API as described by christianklotz
paths:
  /stores:
    get:
      operationId: getFoo
      responses:
        200:
          description: 'ok'
          schema:
            type: array
            items:
              $ref: '#/definitions/Foo'
definitions:
  Foo:
    type: object
    properties:
      arrayHasOmitEmptyTrue:
        type: array
        x-omitempty: true
        items:
          type: string
          enum:
          - some
          - values
      arrayHasOmitEmptyFalse:
        type: array
        x-omitempty: false
        items:
          type: string
          enum:
          - some
          - values
      arrayNoOmitEmpty:
        type: array
        items:
          type: string
          enum:
          - some
          - values
      generalHasOmitEmptyTrue:
        type: string
        x-omitempty: true
      generalHasOmitEmptyFalse:
        type: string
        x-omitempty: false
      generalNoOmitEmpty:
        type: string
      refHasOmitEmptyTrue:
        x-omitempty: true
        $ref: '#/definitions/bar'
      refHasOmitEmptyFalse:
        x-omitempty: false
        $ref: '#/definitions/bar'
      refNoOmitEmpty:
        $ref: '#/definitions/bar'
  bar:
    type: object
    x-nullable: false
    properties:
      prop:
        type: string
```

should generate model like

```go
type Foo struct {
			            	
	// array has omit empty false
	ArrayHasOmitEmptyFalse []string `json:"arrayHasOmitEmptyFalse"`
			            	
	// array has omit empty true
	ArrayHasOmitEmptyTrue []string `json:"arrayHasOmitEmptyTrue,omitempty"`
			            	
	// array no omit empty
	ArrayNoOmitEmpty []string `json:"arrayNoOmitEmpty"`
			            	
	// general has omit empty false
	GeneralHasOmitEmptyFalse string `json:"generalHasOmitEmptyFalse"`
			            	
	// general has omit empty true
	GeneralHasOmitEmptyTrue string `json:"generalHasOmitEmptyTrue,omitempty"`
			            	
	// general no omit empty
	GeneralNoOmitEmpty string `json:"generalNoOmitEmpty,omitempty"`
			            	
	// ref has omit empty false
	RefHasOmitEmptyFalse Bar `json:"refHasOmitEmptyFalse"`
			            	
	// ref has omit empty true
	RefHasOmitEmptyTrue Bar `json:"refHasOmitEmptyTrue,omitempty"`
			            	
	// ref no omit empty
	RefNoOmitEmpty Bar `json:"refNoOmitEmpty,omitempty"`
}
```

### CODE REVIEW

**Comment by zhuangqh** [comment]
@fredbi PTAL thx

**Comment by fredbi** [comment]
Did you pass circle CI and appveyor? I don't see the track of this CI pass

File: docs/use/spec.md [hunk 1]
```diff
@@ -117,4 +117,4 @@ Vendor extension | Description
 `x-go-name`      | the go name of a type
 `x-go-package`   | the go package of a type
 `x-class`        | this is used in conjunction with discriminators to give a full type name
-`x-omitempty`    | this is used with arrays to control presence of omitempty tag to be used by JSON Marshaler
+`x-omitempty`    | this is used to control presence of omitempty tag to be used by JSON Marshaler. if the field is required, this extension don't take effect
```
Comment (fredbi) [inline docs/use/spec.md]: typo: doesn't take effect

File: generator/model_test.go [hunk 2]
```diff
@@ -157,7 +157,7 @@ func TestGenerateModel_SchemaField(t *testing.T) {
 	var gmp GenSchema
 	gmp.Name = "some name"
 	gmp.OriginalName = "some name"
-	gmp.resolvedType = resolvedType{GoType: "string", IsPrimitive: true}
+	gmp.resolvedType = resolvedType{GoType: "string", IsPrimitive: true, IsEmptyOmitted: true}
```
Comment (fredbi) [inline generator/model_test.go]: is this the new default?

Comment (zhuangqh) [reply generator/model_test.go]: yep. By default, x-omitempty is false for array type, true for other.

**Comment by zhuangqh** [comment]
> Did you pass circle CI and appveyor? I don't see the track of this CI pass

@fredbi just scroll down, and you will see it😄

File: generator/model_test.go (hunk 2 above)
Comment (fredbi) [reply generator/model_test.go]: I am puzzled that this does not affect validation code... isn't it strange?

File: generator/moreschemavalidation_fixtures_test.go [hunk 3]
```diff
@@ -5389,7 +5389,7 @@ func initFixture1336() {
 		`	m.nodesField = val`,
 		`func (m *Graph) UnmarshalJSON(raw []byte) error {`,
 		`	var data struct {`,
-		"		Nodes json.RawMessage `json:\"Nodes,omitempty\"`",
+		"		Nodes json.RawMessage `json:\"Nodes\"`",
```
[the remark this answers is not in the document]
Comment (fredbi) [reply generator/moreschemavalidation_fixtures_test.go]: ok

**Comment by fredbi** [comment]
no I don't see the track of our CI passing here. I am expecting 7 checks and I can see 5.

File: generator/model_test.go (hunk 2 above)
Comment (zhuangqh) [reply generator/model_test.go]: The zero value of bool is false. However, our default value is true. 😅

Comment (fredbi) [reply generator/model_test.go]: no I meant, for null values in arrays. this omitempy stuff is supposed to alter validation right?

**Comment by gregmarr** [comment]
@fredbi The checks window is scrollable.

**Comment by zhuangqh** [comment]
> no I don't see the track of our CI passing here. I am expecting 7 checks and I can see 5.

This PR fails to pass 2 checks due to the decrease  of codecov. Generated code: generator/bindata.go reduce the codecov, but I don't have a good solution on it.😅 @fredbi

**Comment by fredbi** [comment]
@zhuangqh apologies for not seeing this scroll!

**Comment by gregmarr** [comment]
@fredbi To be fair, it's not obvious.  I really dislike the trend towards hidden scrollbars.

**Comment by zhuangqh** [comment]
@casualjim Hey, is it possible to cherry-pick this PR to the newest 0.17.0 release. We are benefit from go-swagger and admire your guy's awesome job, but we needs this feature to upgrade from 0.12.0 to 0.17.0. However, we miss the release date which was 5 days ago.😅Thank you so much!

**Pull request merged**